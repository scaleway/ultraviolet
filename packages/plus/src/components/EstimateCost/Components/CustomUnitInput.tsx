'use client'

import { memo, useMemo } from 'react'
import { useEstimateCost } from '../EstimateCostProvider'
import type { Iteration, Units } from '../types'
import { UnitInput } from './UnitInput'

type CustomUnitInputProps = {
  defaultTimeUnit?: Units
  setIteration: (iteration: Iteration) => void
  iteration: Iteration
  timeUnits: Units[]
}

export const CustomUnitInput = memo(
  ({
    defaultTimeUnit = 'hours',
    setIteration,
    iteration,
    timeUnits,
  }: CustomUnitInputProps) => {
    const { locales } = useEstimateCost()

    const options = useMemo(
      () =>
        timeUnits.map(unit => ({
          value: unit,
          label: locales[`estimate.cost.units.${unit}.label`],
        })),
      [timeUnits, locales],
    )

    const defaultOption = useMemo(
      () => options.find(({ value }) => value === defaultTimeUnit),
      [defaultTimeUnit, options],
    )

    return (
      <UnitInput
        name="iteration"
        onChange={inputValue =>
          setIteration({
            unit: iteration.unit,
            value: inputValue,
          })
        }
        onChangeUnitValue={unitValue => {
          setIteration({
            unit: unitValue as Units,
            value: iteration.value,
          })
        }}
        placeholder="0"
        value={iteration.value}
        unitValue={iteration.unit || defaultOption?.value}
        minValue={1}
        size="medium"
        options={options}
      />
    )
  },
)
