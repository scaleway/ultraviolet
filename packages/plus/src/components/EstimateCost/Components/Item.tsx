'use client'

import { css } from '@emotion/react'
import styled from '@emotion/styled'
import { HelpCircleOutlineIcon } from '@ultraviolet/icons'
import { Badge, Stack, Text, Tooltip, zoomIn } from '@ultraviolet/ui'
import type { ComponentProps, ReactNode } from 'react'
import {
  Children,
  cloneElement,
  isValidElement,
  memo,
  useCallback,
  useEffect,
  useId,
  useMemo,
  useState,
} from 'react'
import {
  OverlayRow,
  PriceCell,
  Cell as StyledCell,
  StyledDiv,
  StyledLeftSide,
  StyledTr,
} from '../componentStyle'
import {
  MAX_CELL_WIDTH,
  maximumFractionDigits,
  maximumFractionDigitsLong,
  multiplier,
} from '../constants'
import { useEstimateCost } from '../EstimateCostProvider'
import { calculatePrice } from '../helper'
import { useOverlay } from '../OverlayContext'
import type {
  BareEstimateProduct,
  EstimateProduct,
  Iteration,
  Units,
} from '../types'

const TIME_RELATED_UNIT: Units[] = [
  'seconds',
  'minutes',
  'hours',
  'days',
  'months',
]

const StyledResourceName = styled('div', {
  shouldForwardProp: prop => !['isOverlay', 'animated'].includes(prop),
})<{
  isOverlay: boolean
  animated: boolean
}>`
  text-align: ${({ isOverlay }) => (isOverlay ? 'initial' : 'right')};

  ${({ isOverlay, animated }) =>
    isOverlay
      ? css`
          height: 48px;
          display: flex;
          flex-direction: column;
          -webkit-box-pack: center;
          justify-content: center;
          animation: ${animated ? css`800ms ${zoomIn}` : ''};
        `
      : null}
`

const StyledBadge = styled(Badge)`
  margin-left: ${({ theme }) => theme.space['1']};
  align-self: center;
`

const StyledText = styled(Text)`
  margin-left: ${({ theme }) => theme.space['1']};
`

const MaxWidthText = styled(Text)`
  max-width: 75%;
`

const TextAlignRight = styled(Text)`
  text-align: right;
`

const StyledTooltip = styled(Tooltip)`
  vertical-align: text-top;
`

const StyledPriceCell = styled(StyledCell)`
  ${({ theme }) => PriceCell(theme)}
`

type ExtraProps = {
  itemCallback: (amount: number, isVariant: boolean) => void
  amount: number
  maxAmount: number
  unit: string
}

type ItemProps = {
  amount?: number
  /**
   * Number of items that are free. It will be deducted to price of the item.
   */
  amountFree?: number
  /**
   * if true, zoomIn animation is triggered
   */
  animated?: boolean
  children?: ReactNode
  discount?: number
  discountText?: string
  /**
   * Hide element from overlay
   */
  hideFromOverlay?: boolean
  /**
   * Do not set this prop, it is transferred from parent
   */
  isDefined?: boolean
  /**
   * Do not set this prop, it is transferred from parent
   */
  isFirstElement?: boolean
  /**
   * Do not set this prop, it is transferred from parent
   */
  isLastElement?: boolean
  isPrimaryBackground?: boolean
  /**
   * Do not set this prop, it is transferred from parent
   */
  iteration?: Iteration
  /**
   * String that is displayed on left part of the item, it defines the item
   */
  label?: ReactNode
  /**
   * Changes label typography variant
   */
  labelTextVariant?: ComponentProps<typeof Text>['variant']
  /**
   * Changes label typography prominence
   */
  labelTextProminence?: ComponentProps<typeof Text>['prominence']
  /**
   * If your price has a lot of number after decimal point (ex: 0.0000076) - up 8 fraction digits
   */
  longFractionDigits?: boolean
  /**
   * Create a range price with [amount - maxAmount] values and prices
   */
  maxAmount?: number
  /**
   * Price per month
   */
  monthlyPrice?: number
  /**
   * Remove border bottom of the item
   */
  noBorder?: boolean
  /**
   * If the price of the item is not based on time enable this prop. ex: 5kg of chocolate is same price over 1 month or 1 hour.
   */
  noIteration?: boolean
  noIterationText?: string
  /**
   * Remove the price on the right section of the table
   */
  noPrice?: boolean
  /**
   * Notice to display below the label
   */
  notice?: string
  onAmountChange?: (amount: number) => void
  /**
   * Hourly price for one unit
   */
  price?: number
  priceText?: ReactNode
  productsCallback?: {
    add: (product: EstimateProduct) => void
    remove: (product: BareEstimateProduct) => void
  }
  /**
   * Hide item from overlay if screen width is small
   */
  shouldBeHidden?: boolean
  /**
   * Display a complementary text on the right of the label
   */
  subLabel?: string
  /**
   * Increase left padding of the item like
   */
  tabulation?: number
  /**
   * Text to display in case of not defined value
   */
  textNotDefined?: string
  /**
   * Display near label, an icon with a tooltip that contains your text
   */
  tooltipInfo?: string
  /**
   * Unit of your item, examples: GB, MB, Node, Queries, etc.
   */
  unit?:
    | 'mb'
    | 'gb'
    | 'tb'
    | 'seconds'
    | 'minutes'
    | 'hours'
    | 'days'
    | 'months'
    | 'years'
    // Allow a string for unit but keep autocomplete for the above values
    | (string & NonNullable<unknown>)
  /*
   * To strike through the price
   */
  strikeThrough?: boolean
}

const StyleNoPriceItem = styled(Text)`
  text-align: right;
`

export const Item = memo(
  ({
    discount = 0,
    priceText,
    discountText,
    label,
    tooltipInfo, // Shows an icon with tooltip that contains this text
    subLabel = '', // Usually used for showing amount that is free
    price: basePrice = 0, // Hourly price for one unit
    monthlyPrice = 0, // Price per month
    unit: baseUnit, // Can be GB, MB, Node, Queries, etc.
    amount: currentAmount = 1, // Current number of items
    onAmountChange,
    amountFree = 0, // Amount that is free - offered by company
    maxAmount = 0, // Max amount - used for kubernetes for example
    longFractionDigits = false, // In case price is really long 0.0000076 - up 7 fraction digits
    noIteration = false, // if item is not based on time (ex: download, upload, transfer)
    noIterationText,
    noBorder, // remove the border bottom of the item
    noPrice, // remove the price on right side of the table
    isDefined = true,
    children = null,
    isFirstElement = false,
    isLastElement = false,
    isPrimaryBackground = false,
    productsCallback,
    iteration: receivedIteration, // Object from parent that contains time period (hours, days, months)
    shouldBeHidden = false, // Hide element from overlay if screen width is small
    hideFromOverlay = false, // Hide element from overlay in any case
    textNotDefined, // Text to display in case of not defined value
    animated = false, // if true, zoomIn animation is triggered
    tabulation, // Increase left padding of the item
    labelTextVariant, // To change left cell typography variant
    labelTextProminence, // To change left cell typography prominence
    notice, // To display a gray text below the label
    strikeThrough, // To strike through the price
  }: ItemProps) => {
    const { locales, formatNumber } = useEstimateCost()

    let iteration: Iteration | undefined

    if (noIteration) {
      iteration = {
        ...(receivedIteration ?? { value: 0 }),
        unit: 'hours',
      }
    } else {
      iteration = receivedIteration
    }

    const price = useMemo(() => {
      if (monthlyPrice && basePrice === 0) {
        return monthlyPrice / multiplier.months
      }

      return basePrice
    }, [basePrice, monthlyPrice])

    const unit = useMemo(() => {
      if (!baseUnit) {
        return locales['estimate.cost.units.gb.label']
      }

      return baseUnit
    }, [baseUnit, locales])

    const { isOverlay } = useOverlay()
    const Row = isOverlay ? OverlayRow : StyledTr
    const Cell = isOverlay ? StyledCell.withComponent('div') : StyledCell
    const LeftSide = isOverlay ? 'div' : StyledLeftSide

    const [amount, setAmount] = useState(currentAmount)
    const [isVariant, setIsVariant] = useState(false)

    useEffect(() => setAmount(currentAmount), [setAmount, currentAmount])
    useEffect(() => onAmountChange?.(amount), [onAmountChange, amount])

    const itemCallback = useCallback(
      (localAmount: number, localIsVariant: boolean) => {
        setAmount(localAmount)
        setIsVariant(localIsVariant)
      },
      [setAmount, setIsVariant],
    )

    const id = useId()

    // We remove Item from object list when Iem component unmount to avoid duplicates
    useEffect(
      () => () => productsCallback?.remove({ id }),
      [id, productsCallback],
    )

    useEffect(() => {
      if (!isOverlay) {
        productsCallback?.add({
          amount,
          amountFree,
          discount,
          id,
          isVariant,
          longFractionDigits,
          maxAmount,
          noIteration,
          price,
        })
      }
    }, [
      price,
      discount,
      amount,
      id,
      productsCallback,
      maxAmount,
      noIteration,
      isVariant,
      amountFree,
      isOverlay,
      longFractionDigits,
    ])

    const computedItemPrice = useMemo(
      () =>
        calculatePrice({
          amount,
          amountFree,
          discount,
          price,
          timeAmount: noIteration ? 1 : (iteration?.value ?? 1),
          timeUnit: noIteration ? 'hours' : (iteration?.unit ?? 'hours'),
        }),
      [price, amount, amountFree, iteration, noIteration, discount],
    )

    const computedMaxItemPrice = useMemo(
      () =>
        calculatePrice({
          amount: maxAmount,
          amountFree,
          discount,
          price,
          timeAmount: noIteration ? 1 : (iteration?.value ?? 1),
          timeUnit: noIteration ? 'hours' : (iteration?.unit ?? 'hours'),
        }),
      [price, maxAmount, amountFree, iteration, noIteration, discount],
    )

    const formatMaximumFractionDigits = useMemo(() => {
      if (!iteration?.unit) {
        return undefined
      }

      return longFractionDigits
        ? maximumFractionDigitsLong[iteration?.unit]
        : maximumFractionDigits[iteration?.unit]
    }, [iteration?.unit, longFractionDigits])

    return (
      <Row
        hideFromOverlay={hideFromOverlay}
        isFirstElement={isFirstElement}
        shouldBeHidden={shouldBeHidden}
      >
        <Cell
          hasBorder={!isLastElement && !noBorder && !isOverlay}
          tabulation={tabulation}
          width={!isOverlay ? MAX_CELL_WIDTH : 'inherit'}
        >
          <LeftSide>
            <Stack>
              <Stack direction="row">
                <Text
                  as="p"
                  prominence={labelTextProminence ?? 'default'}
                  variant={labelTextVariant ?? 'body'}
                >
                  {label}
                </Text>
                {tooltipInfo ? (
                  <StyledDiv>
                    <StyledTooltip text={tooltipInfo}>
                      <HelpCircleOutlineIcon size="medium" />
                    </StyledTooltip>
                  </StyledDiv>
                ) : null}
                {subLabel && !isOverlay ? (
                  <StyledText as="p" italic sentiment="primary" variant="body">
                    {subLabel}
                  </StyledText>
                ) : null}
                {discount > 0 && discountText ? (
                  <StyledBadge
                    prominence="strong"
                    sentiment="warning"
                    size="small"
                  >
                    {discountText}
                  </StyledBadge>
                ) : null}
              </Stack>
              {notice ? (
                <MaxWidthText as="p" prominence="weak" variant="caption">
                  {notice}
                </MaxWidthText>
              ) : null}
            </Stack>
            <StyledResourceName animated={animated} isOverlay={isOverlay}>
              {isDefined
                ? Children.map(children, child =>
                    isValidElement<ExtraProps>(child)
                      ? cloneElement(child, {
                          amount,
                          itemCallback,
                          maxAmount,
                          unit,
                        })
                      : null,
                  )
                : textNotDefined || locales['estimate.cost.notDefined']}
            </StyledResourceName>
          </LeftSide>
        </Cell>
        {!isOverlay ? (
          <StyledPriceCell
            hasBorder={!isLastElement && !noBorder}
            primary={isPrimaryBackground}
          >
            {!noPrice ? (
              <>
                <StyleNoPriceItem
                  as="p"
                  prominence={
                    computedItemPrice === 0 && computedMaxItemPrice === 0
                      ? 'weak'
                      : 'default'
                  }
                  sentiment={
                    computedItemPrice === 0 && computedMaxItemPrice === 0
                      ? 'neutral'
                      : 'primary'
                  }
                  strikeThrough={strikeThrough}
                  variant={noIterationText ? 'headingSmall' : 'bodyStrong'}
                >
                  {priceText}
                  {!priceText
                    ? formatNumber(computedItemPrice, {
                        maximumFractionDigits: formatMaximumFractionDigits,
                      })
                    : null}
                  {noIterationText ? (
                    <Text as="span" sentiment="primary" variant="bodySmall">
                      /{noIterationText}
                    </Text>
                  ) : null}
                  {!priceText && computedMaxItemPrice > 0
                    ? ` - ${formatNumber(computedMaxItemPrice, {
                        maximumFractionDigits: formatMaximumFractionDigits,
                      })}`
                    : null}
                </StyleNoPriceItem>
                {(amount - amountFree !== 1 && computedItemPrice > 0) ||
                (maxAmount > 0 && computedMaxItemPrice > 0) ? (
                  <TextAlignRight as="p" variant="body">
                    {formatNumber(
                      calculatePrice({
                        amount: 1,
                        discount,
                        price,
                        timeAmount: 1,
                        timeUnit: 'hours',
                      }),
                      {
                        maximumFractionDigits: longFractionDigits
                          ? maximumFractionDigitsLong.hours
                          : maximumFractionDigits.hours,
                      },
                    )}
                    {TIME_RELATED_UNIT.includes(unit as Units)
                      ? locales[
                          `estimate.cost.units.${unit as Units}.label`
                        ].toLowerCase()
                      : `/${unit}`}
                    {!noIteration
                      ? `/${locales[
                          'estimate.cost.units.hours.label'
                        ].toLowerCase()}`
                      : null}
                  </TextAlignRight>
                ) : null}
              </>
            ) : null}
          </StyledPriceCell>
        ) : null}
      </Row>
    )
  },
)
