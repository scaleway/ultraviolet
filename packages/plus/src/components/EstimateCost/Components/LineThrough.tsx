'use client'

import type { ComponentProps } from 'react'
import { estimateCostLineThrough } from './components.css'

type LineThroughProps = ComponentProps<'span'> & {
  isActive?: boolean
}

export const LineThrough = ({ isActive, ...props }: LineThroughProps) => (
  <span
    className={`${props.className ? `${props.className} ` : ''} ${isActive ? estimateCostLineThrough : ''}`}
    {...props}
  />
)
