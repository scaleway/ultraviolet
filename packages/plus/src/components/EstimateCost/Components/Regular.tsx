'use client'

import { css } from '@emotion/react'
import styled from '@emotion/styled'
import type { ReactNode } from 'react'
import { memo } from 'react'
import { useOverlay } from '../OverlayContext'

type RegularProps = {
  variant?: 'normal' | 'small' | 'big' | 'capitalized'
  isDisabledOnOverlay?: boolean
  children?: ReactNode
  className?: string
}

const StyledRegular = styled('div', {
  shouldForwardProp: prop => !['variant', 'isOverlay'].includes(prop),
})<{
  isOverlay: boolean
  variant: 'normal' | 'small' | 'big' | 'capitalized'
}>`
  display: ${({ isOverlay }) => (isOverlay ? 'flex' : 'inline-flex')};
  max-width: 500px;
  align-items: center;
  font-size: 16px;
  color: ${({ theme }) => theme.colors.neutral.textStrong};
  margin-right: 4px;

  ${({ theme, variant }) =>
    variant === 'small' &&
    css`
      display: block;
      font-size: 14px;
      line-height: 8px;
      color: ${theme.colors.neutral.text};
    `};
`

export const Regular = memo(
  ({
    variant = 'normal',
    isDisabledOnOverlay = false,
    children = null,
    className,
  }: RegularProps) => {
    const { isOverlay } = useOverlay()

    return !isDisabledOnOverlay || !isOverlay ? (
      <StyledRegular
        className={className}
        variant={variant}
        isOverlay={isOverlay}
      >
        {children}
      </StyledRegular>
    ) : null
  },
)
