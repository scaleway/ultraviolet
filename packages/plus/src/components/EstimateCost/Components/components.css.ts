import { theme } from '@ultraviolet/themes'
import { zoomInVanillaExtract } from '@ultraviolet/ui'
import { createVar, style } from '@vanilla-extract/css'
import { recipe } from '@vanilla-extract/recipes'

export const maxWidthTextVar = createVar()

export const maxWidthText = style({
  maxWidth: maxWidthTextVar,
})

export const estimateCostImage = style({
  width: 15,
  marginRight: theme.space[1],
})

export const estimateCostTr = style({})

export const styledDiv = style({ marginLeft: theme.space['0.5'] })

export const estimateCostLeftSide = style({
  display: 'flex',
  flexDirection: 'row',
  WebkitBoxPack: 'justify',
  justifyContent: 'space-between',
  alignItems: 'center',
  minHeight: theme.sizing[700],
  paddingTop: theme.space[1],
  paddingBottom: theme.space[1],
})

export const estimateCostItemResourceName = recipe({
  base: {
    height: theme.sizing[600],
    display: 'flex',
    flexDirection: 'column',
    WebkitBoxPack: 'center',
    justifyContent: 'center',
  },
  variants: {
    animated: {
      true: {
        animation: `800ms ${zoomInVanillaExtract}`,
      },
    },
  },
  defaultVariants: {
    animated: false,
  },
})

export const estimateCostResourceName = recipe({
  variants: {
    isOverlay: {
      true: {
        textAlign: 'initial',
        height: theme.sizing[600],
        display: 'flex',
        flexDirection: 'column',
        WebkitBoxPack: 'center',
        justifyContent: 'center',
      },
      false: {
        textAlign: 'right',
      },
    },
    isAnimated: {
      true: {
        animation: `${zoomInVanillaExtract} 800ms`,
      },
    },
  },
  defaultVariants: {
    isAnimated: false,
    isOverlay: false,
  },
})

export const estimateCostBadgeItem = style({
  marginLeft: theme.space[1],
  alignSelf: 'center',
})

export const estimateCostTextItem = style({
  marginLeft: theme.space[1],
})

export const estimateCostMaxWidthText = style({
  maxWidth: '75%',
})

export const estimateCostTooltip = style({ verticalAlign: 'text-top' })

export const estimateCostLineThrough = style({
  textDecorationLine: 'line-through',
  textDecorationColor: theme.colors.warning.border,
})

export const estimateCostRegular = recipe({
  base: {
    maxWidth: 500,
    alignItems: 'center',
    fontSize: 16,
    color: theme.colors.neutral.textStrong,
    marginRight: theme.space['0.5'],
  },
  variants: {
    isOverlay: {
      true: {
        display: 'flex',
      },
      false: {
        display: 'inline-flex',
      },
    },
    variant: {
      normal: {},
      big: {},
      capitalized: {},
      small: {
        display: 'block',
        fontSize: 14,
        lineHeight: 8,
        color: theme.colors.neutral.text,
      },
    },
  },
  defaultVariants: {
    isOverlay: false,
    variant: 'normal',
  },
})

export const estimateCostStrong = recipe({
  base: {
    display: 'inline-flex',
    alignItems: 'center',
    color: theme.colors.neutral.textStrong,
    fontWeight: 500,
    marginRight: 4,
  },
  variants: {
    variant: {
      normal: {
        fontSize: 16,
      },
      big: {
        fontSize: 24,
      },
      capitalized: {
        fontSize: 16,
        textTransform: 'capitalize',
      },
      small: {
        fontSize: 16,
      },
    },
  },
  defaultVariants: {
    variant: 'normal',
  },
})

export const estimateCostNumberInput = style({
  MozAppearance: 'textfield',
  selectors: {
    '&::-webkit-inner-spin-button, &::webkit-outer-spin-button': {
      WebkitAppearance: 'none',
      margin: 0,
    },
  },
})
