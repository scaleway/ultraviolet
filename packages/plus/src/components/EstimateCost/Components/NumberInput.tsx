'use client'

import { NumberInput as NumberInputUV } from '@ultraviolet/ui'
import { useEffect, useState } from 'react'
import { ItemResourceName } from '../componentStyle'
import { useOverlay } from '../OverlayContext'
import { Regular } from './Regular'

type NumberInputProps = {
  amount?: number
  itemCallback?: (amount?: number | null, isVariant?: boolean) => void
  getAmountValue?: (amount?: number | null) => void
  minValue?: number
  maxValue?: number
  controls?: boolean
}

export const NumberInput = ({
  amount,
  minValue = 0,
  maxValue = 100,
  getAmountValue,
  itemCallback,
  controls = true,
}: NumberInputProps) => {
  const { isOverlay } = useOverlay()
  const [value, setValue] = useState<number | undefined | null>(amount)

  useEffect(() => {
    getAmountValue?.(amount)
  }, [getAmountValue, amount])

  return isOverlay ? (
    <ItemResourceName animated={false}>
      <Regular>{amount}</Regular>
    </ItemResourceName>
  ) : (
    <NumberInputUV
      min={minValue}
      max={maxValue}
      size="small"
      onChange={newValue => {
        setValue(newValue)
        itemCallback?.(newValue, true)
        getAmountValue?.(newValue)
      }}
      value={value}
      controls={controls}
    />
  )
}
