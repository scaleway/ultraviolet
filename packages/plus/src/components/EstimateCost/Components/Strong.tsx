'use client'

import styled from '@emotion/styled'
import { type ReactNode, memo } from 'react'
import { Regular } from './Regular'

type StrongProps = {
  variant?: 'normal' | 'small' | 'big' | 'capitalized'
  children?: ReactNode
  isDisabledOnOverlay?: boolean
}

export const StyledStrong = styled(Regular, {
  shouldForwardProp: prop => !['variant'].includes(prop),
})<StrongProps>`
  display: inline-flex;
  align-items: center;
  font-size: ${({ variant }) => (variant === 'big' ? 24 : 16)}px;
  ${({ variant }) =>
    variant === 'capitalized' ? `text-transform: capitalize;` : ''}
  color: ${({ theme }) => theme.colors.neutral.textStrong};
  font-weight: 500;
  margin-right: 4px;
`

export const Strong = memo(
  ({
    variant = 'normal',
    children = null,
    isDisabledOnOverlay = false,
  }: StrongProps) => (
    <StyledStrong variant={variant} isDisabledOnOverlay={isDisabledOnOverlay}>
      {children}
    </StyledStrong>
  ),
)
