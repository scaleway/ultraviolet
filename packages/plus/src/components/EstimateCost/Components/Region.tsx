'use client'

import styled from '@emotion/styled'
import type { ComponentProps, ReactNode } from 'react'
import { memo } from 'react'
import { useEstimateCost } from '../EstimateCostProvider'
import type { BareEstimateProduct, EstimateProduct, Iteration } from '../types'
import { Item } from './Item'
import { Strong } from './Strong'

const StyledImage = styled.img`
  width: 15px;
  margin-right: ${({ theme }) => theme.space['1']};
`

type RegionProps = {
  shouldBeHidden?: boolean
  priceText?: ReactNode
  animated?: boolean
  isFirstElement?: boolean
  isLastElement?: boolean
  productsCallback?: {
    add: (product: EstimateProduct) => void
    remove: (product: BareEstimateProduct) => void
  }
  iteration?: Iteration
  discount?: number
  label: string
  image: string
  noBorder?: boolean
  noPrice?: boolean
} & Pick<ComponentProps<typeof Item>, 'hideFromOverlay'>

export const Region = memo(
  ({
    label,
    image,
    shouldBeHidden = false,
    priceText,
    animated = false,
    isFirstElement,
    isLastElement,
    productsCallback,
    iteration,
    discount,
    noBorder,
    noPrice,
    hideFromOverlay,
  }: RegionProps) => {
    const { locales } = useEstimateCost()

    return (
      <Item
        label={locales['estimate.cost.region.label']}
        shouldBeHidden={shouldBeHidden}
        priceText={priceText}
        animated={animated}
        isFirstElement={isFirstElement}
        isLastElement={isLastElement}
        productsCallback={productsCallback}
        iteration={iteration}
        discount={discount}
        noBorder={noBorder}
        noPrice={noPrice}
        hideFromOverlay={hideFromOverlay}
      >
        <Strong>
          <StyledImage alt={label} src={image} />
          {label}
        </Strong>
      </Item>
    )
  },
)
