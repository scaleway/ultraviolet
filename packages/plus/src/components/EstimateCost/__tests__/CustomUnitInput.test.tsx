import { screen, waitFor } from '@testing-library/react'
import { userEvent } from '@testing-library/user-event'
import { renderWithTheme, shouldMatchEmotionSnapshot } from '@utils/test'
import {
  resetIntersectionMocking,
  setupIntersectionMocking,
} from 'react-intersection-observer/test-utils'
import { afterEach, beforeEach, describe, expect, test, vi } from 'vitest'
import { CustomUnitInput } from '../Components/CustomUnitInput'

describe('EstimateCost - CustomUnitInput', () => {
  beforeEach(() => {
    setupIntersectionMocking(vi.fn)
  })
  afterEach(() => {
    resetIntersectionMocking()
  })
  test('render default values', () =>
    shouldMatchEmotionSnapshot(
      <CustomUnitInput
        iteration={{ unit: 'hours', value: 1 }}
        setIteration={() => {}}
        timeUnits={['seconds', 'minutes', 'hours', 'days', 'months']}
      />,
    ))

  test('render and trigger on blur when leaving input empty', async () => {
    renderWithTheme(
      <CustomUnitInput
        iteration={{ unit: 'hours', value: 1 }}
        setIteration={() => {}}
        timeUnits={['seconds', 'minutes', 'hours', 'days', 'months']}
      />,
    )

    const input = screen.getByRole<HTMLInputElement>('spinbutton')
    await waitFor(() => expect(input.value).toBe('1'))
    await userEvent.click(input)
    await userEvent.type(input, '{ArrowLeft}{Backspace}0')
    await userEvent.tab()

    await waitFor(() => expect(input.value).toBe('1'))
  })
})
