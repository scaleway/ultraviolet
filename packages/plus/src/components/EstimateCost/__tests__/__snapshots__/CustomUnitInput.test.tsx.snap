// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`estimateCost - CustomUnitInput > render default values 1`] = `
<DocumentFragment>
  .emotion-0 {
  width: 100%;
}

.emotion-2 {
  display: grid;
  width: 100%;
  gap: 0.5rem;
  grid-template-columns: 1fr auto;
  padding: 0.5rem;
  padding-right: 0;
  padding-left: 1rem;
  cursor: pointer;
  box-shadow: none;
  background: #ffffff;
  border-radius: 0.25rem;
  width: 100%;
}

.emotion-2[data-size='small'] {
  height: 2rem;
  padding-left: 0.5rem;
}

.emotion-2[data-size='medium'] {
  height: 2.5rem;
}

.emotion-2[data-size='large'] {
  height: 3rem;
}

.emotion-2[data-state='neutral'] {
  border: 1px solid #d9dadd;
}

.emotion-2[data-state='neutral']:not([data-disabled="true"]):not([data-readonly="true"]):active {
  border-color: #792dd4;
  box-shadow: 0px 0px 0px 3px #8c40ef40;
}

.emotion-2[data-state='neutral']:not([data-disabled='true']):hover {
  border-color: #792dd4;
  outline: none;
}

.emotion-2[data-state='neutral']:not([data-disabled='true']):focus-visible {
  outline: 5px auto Highlight;
  outline: 5px auto -webkit-focus-ring-color;
}

.emotion-2[data-state='neutral'][data-dropdownvisible='true'] {
  border-color: #792dd4;
}

.emotion-2[data-state='success'] {
  border: 1px solid #22674e;
}

.emotion-2[data-state='success']:not([data-disabled="true"]):not([data-readonly="true"]):active {
  border-color: #1b533f;
  box-shadow: 0px 0px 0px 3px #45d19f40;
}

.emotion-2[data-state='success'][data-dropdownvisible='true'] {
  border-color: #1b533f;
}

.emotion-2[data-state='danger'] {
  border: 1px solid #b3144d;
}

.emotion-2[data-state='danger']:not([data-disabled="true"]):not([data-readonly="true"]):active {
  border-color: #92103f;
  box-shadow: 0px 0px 0px 3px #f91b6c40;
}

.emotion-2[data-state='danger'][data-dropdownvisible='true'] {
  border-color: #92103f;
}

.emotion-2:not([data-disabled='true']):not([data-readonly]):hover {
  border-color: #8c40ef;
}

.emotion-2[data-readonly='true'] {
  background: #f9f9fa;
  border-color: #d9dadd;
  cursor: default;
}

.emotion-2[data-disabled='true'] {
  background: #f3f3f4;
  border-color: #e9eaeb;
  cursor: not-allowed;
}

.emotion-4 {
  text-overflow: ellipsis;
  overflow: hidden;
  white-space: nowrap;
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
}

.emotion-6 {
  padding-right: 1rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

<div
    data-testid="testing"
  >
    <div
      class="uv_toi52u0 uv_toi52u31 uv_toi52u2d uv_toi52u7d uv_toi52u1v uv_toi52u5d"
      style="--uv_4k0ekn0: 100%;"
    >
      <div
        class="uv_gwf0fld uv_gwf0fla uv_gwf0fl5 uv_x6hyh50 "
        style="--uv_3mg8xe0: auto auto; --uv_3mg8xe1: auto auto; --uv_3mg8xe2: auto auto; --uv_3mg8xe3: auto auto; --uv_3mg8xe4: auto auto; --uv_3mg8xe5: auto auto;"
      >
        <div
          class="uv_gwf0flb"
          id="input-field"
        >
          <input
            aria-invalid="false"
            class="uv_gwf0fl3 uv_gwf0fl1"
            data-testid="unit-input"
            id="«r0»"
            max="9007199254740991"
            min="1"
            name="iteration-value"
            placeholder="0"
            step="1"
            type="number"
            value="1"
          />
        </div>
        <div
          aria-label="iteration-unit"
          class="uv_gwf0flg uv_gwf0flf emotion-0 emotion-1"
          style="--uv_gwf0fl0: 100%;"
        >
          <div
            aria-controls="«r2»"
            aria-describedby="«r2»"
            class="uv_w40vpo9 uv_w40vpob"
            tabindex="-1"
          >
            <div
              class="uv_toi52u0 uv_toi52u31 uv_toi52u2d uv_toi52u7d uv_toi52u1v uv_toi52u5d"
            >
              <div
                aria-controls="«r2»"
                aria-expanded="false"
                class="emotion-2 emotion-3"
                data-disabled="false"
                data-dropdownvisible="false"
                data-readonly="false"
                data-size="medium"
                data-state="neutral"
                data-testid="select-input-iteration-unit"
                id="unit"
                role="combobox"
                tabindex="0"
              >
                <span
                  class="emotion-4 emotion-5 uv_m4c9ow0 uv_m4c9ow2 uv_m4c9ow4 uv_m4c9ow6 uv_m4c9ow8 uv_m4c9owb uv_m4c9owi uv_m4c9own uv_m4c9ow1b uv_m4c9ow20 uv_m4c9ow3c"
                >
                  Hour(s)
                </span>
                <div
                  class="emotion-6 emotion-7 uv_toi52u0 uv_toi52u3d uv_toi52u2j uv_toi52u7d uv_toi52u7 uv_toi52u5d"
                >
                  <svg
                    aria-label="show dropdown"
                    class="styles__1sbwqkz0 styles_prominence_default__1sbwqkzg styles_disabled_false__1sbwqkzf styles_sentiment_neutral__1sbwqkz3 styles_size_small__1sbwqkza styles_undefined_compound_34__1sbwqkz1i"
                    viewBox="0 0 16 16"
                  >
                    <path
                      clip-rule="evenodd"
                      d="M4.22 6.22a.75.75 0 0 1 1.06 0L8 8.94l2.72-2.72a.75.75 0 1 1 1.06 1.06l-3.25 3.25a.75.75 0 0 1-1.06 0L4.22 7.28a.75.75 0 0 1 0-1.06"
                      fill-rule="evenodd"
                    />
                  </svg>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
